src/OrchardCore.Modules/OrchardCore.AutoSetup/AutoSetupMiddleware.cs

```diff
@@ -0,0 +1,212 @@
+        /// </summary>
+        private readonly IShellSettingsManager _shellSettingsManager;
+
+        /// <summary>
+        /// The auto-setup options.
+        /// </summary>
+        private readonly IOptions<AutoSetupOptions> _setupOptions;
+
+        private readonly IShellHost _shellHost;
+
+        /// <summary>
+        /// The _logger.
+        /// </summary>
+        private readonly ILogger<AutoSetupMiddleware> _logger;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AutoSetupMiddleware"/> class.
+        /// </summary>
+        /// <param name="next"> The next middleware in the execution pipeline. </param>
+        /// <param name="shellSettingsManager">The Shell settings manager</param>
+        /// <param name="setupOptions"> The auto-setup Options. </param>
+        /// <param name="shellHost"></param>
+        /// <param name="logger"> The logger. </param>
+        public AutoSetupMiddleware(
+            RequestDelegate next,
+            IShellSettingsManager shellSettingsManager,
+            IOptions<AutoSetupOptions> setupOptions,
+            IShellHost shellHost,
+            ILogger<AutoSetupMiddleware> logger)
+        {
+            _logger = logger;
+            _next = next;
+            _shellSettingsManager = shellSettingsManager;
+            _setupOptions = setupOptions;
+            _shellHost = shellHost;
+        }
+
+        /// <summary>
+        /// The middleware auto-setup invoke.
+        /// </summary>
+        /// <param name="httpContext">
+        /// The http context.
+        /// </param>
+        /// <returns>
+        /// The <see cref="Task"/>.
+        /// </returns>
+        public async Task InvokeAsync(HttpContext httpContext)
+        {
+            var setupService = httpContext.RequestServices.GetRequiredService<ISetupService>();
+            var shellSettings = httpContext.RequestServices.GetRequiredService<ShellSettings>();
+
+            if (shellSettings.State == TenantState.Uninitialized)
+            {
+                var setupTenant = _setupOptions.Value.Tenants
+                    .FirstOrDefault(options => string.Equals(shellSettings.Name, options.ShellName));
+
+                if (setupTenant != null)
+                {
+                    if (await SetupTenantAsync(setupService, setupTenant, shellSettings))
+                    {
+                        if (setupTenant.IsDefault)
+                        {
+                            // Create the rest of the Shells for further on demand setup
+                            foreach (var tenant in _setupOptions.Value.Tenants)
+                            {
+                                if (setupTenant != tenant)
+                                {
+                                    await CreateTenantSettingsAsync(tenant);
+                                }
+                            }
+                        }
+                    }
+
+                    var redirectUrl = "/";
```

Comment:
Did you try my suggestion? The pathBase contains an eventual virtual folder, and at this point it already contains the tenant prefix (the default tenant may have one prefix).

    httpContext.Response.Redirect(httpContext.Request.PathBase)

Reply:
>Did you try my suggestion?

Yes, as you mentioned below context.Request.PathBase can be empty in this case I have to redirect to /.  
Plus since we have `AutoSetupOptions.AutoSetupPath` option the Request Url can be different from Tenant Url in this case I have to redirect to Tenant's Url